### PR TITLE
Allow early stop based on walltime since improvement

### PR DESCRIPTION
**Pull request by jppgks**
This PR:
- stores the Posix timestamp of the last improvement in ProgressTracker
- adds a convenience callback implementing `should_early_stop` that returns True if the given timeout has been reached and no improvement has been made for the given early stopping steps

### CODE REVIEW

**Review by tgaddair — APPROVED** [review]

**Review by justinxzhao — APPROVED** [review]
LGTM!

File: ludwig/utils/trainer_utils.py [hunk 1]
```diff
@@ -163,6 +171,39 @@ def log_metrics(self):
...
+        self.early_stopping_steps = early_stopping_steps
+
+    def should_early_stop(self, trainer, progress_tracker, is_coordinator):
+        """
+        Returns:
+            True if the timeout has been reached and no improvement has been made for the given early stopping steps.
+        """
+        # Time since last improvement > timeout
+        timed_out = (
+            datetime.now() - datetime.fromtimestamp(progress_tracker.last_improvement_timestamp) > self.timeout_delta
```
Comment (tgaddair) [inline ludwig/utils/trainer_utils.py:196]: Actually, something important here: how does this work with resume from previous training run? We may want to make sure this value is reset explicitly when the user is resuming from a previous training run.

**Review by tgaddair — CHANGES_REQUESTED** [review]
Can we also wire this into the config alongside `early_stop`, something like:

```
trainer:
    early_stop_timeout_s = 3600
```

File: ludwig/utils/trainer_utils.py (hunk 1 above)
Comment (jppgks) [reply ludwig/utils/trainer_utils.py:278]: @tgaddair where does the ProgressTracker get loaded/initialized when resuming from a previous training run?

Comment (justinxzhao) [reply ludwig/utils/trainer_utils.py:277]: `ProgressTracker` gets loaded in the trainer, here:

https://github.com/ludwig-ai/ludwig/blob/master/ludwig/trainers/trainer.py#L683

File: ludwig/utils/trainer_utils.py [hunk 2]
```diff
@@ -244,6 +255,40 @@ def log_metrics(self):
...
+        self.timeout_delta = timedelta(seconds=timeout_s)
+        self.early_stopping_steps = early_stopping_steps
+
+    def should_early_stop(self, trainer, progress_tracker, is_coordinator):
+        """
+        Returns:
+            True if the timeout has been reached and no improvement has been made for the given early stopping steps.
+        """
+        # Time since last improvement > timeout
+        timed_out = (
```
Comment (tgaddair) [inline ludwig/utils/trainer_utils.py:280]: @jppgks are you testing this comparison? I don't think you can compare the difference between two datetimes against a float:

```
a = datetime.now()
b = datetime.now()
b-a > 10
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: '>' not supported between instances of 'datetime.timedelta' and 'int'
```

I think you need `total_seconds()`. Also, let's make sure we're testing this.

Comment (tgaddair) [reply ludwig/utils/trainer_utils.py:280]: Ah, nevermind, see you're using a `timedelta` here.

**Pull request closed without merging**